pkg/server/secrets-encrypt.go

```diff
@@ -0,0 +1,489 @@
+package server
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"text/tabwriter"
+	"time"
+
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/wrangler/pkg/generated/controllers/core"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiserverconfigv1 "k8s.io/apiserver/pkg/apis/config/v1"
+)
+
+const (
+	EncryptionStart     string = "start"
+	EncryptionPrepare   string = "prepare"
+	EncryptionRotate    string = "rotate"
+	EncryptionReencrypt string = "reencrypt"
+)
+
+const aescbcKeySize = 32
+
+type EncryptionState struct {
+	Stage      string `json:"stage"`
+	CurrentKey apiserverconfigv1.Key
+}
+
+type EncryptionRequest struct {
+	Stage  string `json:"stage,omitempty"`
+	Toggle bool   `json:"toggle,omitempty"`
+	Force  bool   `json:"force"`
```

Comment:
Does toggle toggle it upon receipt of the request, or does it enable it if true? Lets be more declarative about it. Also, using omitempty for non-nilable types gives weird results and should be avoided.

If we make the RotationStep a struct instead of a string, can we use it to both report the current status, and move the rotation to the next step?

```suggestion
type SecretsEncryptionStatus struct {
	RotationStep  string `json:"rotationStep"`
	Enabled bool   `json:"enabled"`
	Force  bool   `json:"force"`
```

Reply:
Reworded the toggle to be enable... it is not a toggle, it is an explicit enable/disable(i.e. not enable).